### PR TITLE
Get rid of need for -T /dev/null with picolibc

### LINKED ISSUE #60040: Building with `CONFIG_PICOLIBC` is broken on Windows due to use of `/dev/null`
See below:

```
C:\Users\carles\src\zephyr\zephyr>west build -b nrf52840dk_nrf52840 samples\hello_world -- -DCONFIG_PICOLIBC=y
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: C:/Users/carles/src/zephyr/zephyr/samples/hello_world
-- CMake version: 3.24.1
-- Found Python3: C:/Python310/python.exe (found suitable exact version "3.10.5") found components: Interpreter
-- Cache files will be written to: C:/Users/carles/src/zephyr/zephyr/.cache
-- Zephyr version: 3.4.99 (C:/Users/carles/src/zephyr/zephyr)
-- Found west (found suitable version "1.0.0", minimum required is "0.14.0")
-- Board: nrf52840dk_nrf52840
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.1 (C:/Users/carles/bin/zephyr-sdk-0.16.1)
-- Found toolchain: zephyr 0.16.1 (C:/Users/carles/bin/zephyr-sdk-0.16.1)
-- Found Dtc: C:/ProgramData/chocolatey/bin/dtc.exe (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: C:/Users/carles/src/zephyr/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Generated zephyr.dts: C:/Users/carles/src/zephyr/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/Users/carles/src/zephyr/zephyr/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: C:/Users/carles/src/zephyr/zephyr/build/zephyr/dts.cmake
Parsing C:/Users/carles/src/zephyr/zephyr/Kconfig
Loaded configuration 'C:/Users/carles/src/zephyr/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration 'C:/Users/carles/src/zephyr/zephyr/samples/hello_world/prj.conf'
Merged configuration 'C:/Users/carles/src/zephyr/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to 'C:/Users/carles/src/zephyr/zephyr/build/zephyr/.config'
Kconfig header saved to 'C:/Users/carles/src/zephyr/zephyr/build/zephyr/include/generated/autoconf.h'
-- Found GnuLd: c:/users/carles/bin/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Users/carles/bin/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc.exe
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/carles/src/zephyr/zephyr/build
-- west build: building application
[1/139] Generating include/generated/version.h
-- Zephyr version: 3.4.99 (C:/Users/carles/src/zephyr/zephyr), build: zephyr-v3.4.0-685-gc3dc01ddf3a5
[134/139] Linking C executable zephyr\zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map C:/Users/carles/src/zephyr/zephyr/build/zephyr/zephyr_pre0.map
cmd.exe /C "cd . && C:\Users\carles\bin\zephyr-sdk-0.16.1\arm-zephyr-eabi\bin\arm-zephyr-eabi-gcc.exe  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr\zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=C:/Users/carles/src/zephyr/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/soc/arm/common/cortex_m/libsoc__arm__common__cortex_m.a  zephyr/soc/arm/nordic_nrf/nrf52/libsoc__arm__nordic_nrf__nrf52.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  modules/hal_nordic/nrfx/libmodules__hal_nordic__nrfx.a  modules/segger/libmodules__segger.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"c:/users/carles/bin/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/thumb/v7e-m/nofp"  -LC:/Users/carles/src/zephyr/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -mcpu=cortex-m4  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_INTEGER_PRINTF_SCANF  -T/dev/null  --specs=picolibc.specs  -lc  -lgcc && cmd.exe /C "cd /D C:\Users\carles\src\zephyr\zephyr\build\zephyr && "C:\Program Files\CMake\bin\cmake.exe" -E true""
c:/users/carles/bin/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe: cannot open linker script file /dev/null: No such file or directory
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\carles\src\zephyr\zephyr\build'
```

### PR DESCRIPTION
**Pull request by keith-packard**
Picolibc's linker flags includes `-T /dev/null`, which won't work on non-POSIX systems. This is used because the picolibc gcc .specs file looks to see if the linker command line includes a linker script. If it doesn't, the specs file inserts the picolibc linker script. This allows simple applications as are often used in build scripts to not need a linker script of their own. 

Zephyr provides a linker script itself, but because clang doesn't support the `-T` option, it used `-Wl,-T` to pass the script file name through the compiler driver to the linker. This meant that picolibc's heuristic for detecting whether a script file has been provided fails and the picolibc linker script gets added to the linker command line. This was kludged around by adding `-T /dev/null` in the picolibc zephyr cmake bits.

Fix this by using `-T` for gcc while leaving other compilers continuing to use `-Wl,-T` when specifying the Zephyr linker script. This allows us to remove the `-T /dev/null` from the picolibc bits.

This should be a reasonable replacement for #60042 
Fixes: #60040

### CODE REVIEW

**Review by cfriedt — APPROVED** [review]

File: cmake/compiler/gcc/compiler_flags.cmake [hunk 1]
```diff
@@ -182,6 +182,9 @@ set_compiler_property(PROPERTY debug -g)
 # Flags to save temporary object files
 set_compiler_property(PROPERTY save_temps -save-temps=obj)
 
+# Flag to specify linker script
+set_compiler_property(PROPERTY linker_script -T)
```
Comment (tejlmand) [inline cmake/compiler/gcc/compiler_flags.cmake:186]: should go to the `linker/ld/gcc` front-end flags here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/linker/ld/gcc/linker_flags.cmake

File: cmake/compiler/clang/compiler_flags.cmake [hunk 2]
```diff
@@ -23,6 +23,9 @@ set_compiler_property(PROPERTY diagnostic -fcolor-diagnostics)
 # clang flag to save temporary object files
 set_compiler_property(PROPERTY save_temps -save-temps)
 
+# clang doesn't handle the -T flag
+set_compiler_property(PROPERTY linker_script -Wl,-T)
```
Comment (tejlmand) [inline cmake/compiler/clang/compiler_flags.cmake:27]: let's have the `-Wl,-T` as the default ld linker script setting here (which can the be overruled by the gcc ld front-end):
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/linker/ld/linker_flags.cmake

as:
```
set_property(TARGET linker PROPERTY linker_script "${LINKERFLAGPREFIX},-T")
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -485,8 +485,9 @@ if(CONFIG_USERSPACE)
 endif()
 
 get_property(TOPT GLOBAL PROPERTY TOPT)
-set_ifndef(  TOPT -Wl,-T) # clang doesn't pick -T for some reason and complains,
-                          # while -Wl,-T works for both, gcc and clang
+get_property(COMPILER_TOPT TARGET compiler PROPERTY linker_script)
```
Comment (tejlmand) [inline CMakeLists.txt:488]: combined with other comments, this should be on linker target.
```suggestion
get_property(COMPILER_TOPT TARGET linker PROPERTY linker_script)
```

**Review by tejlmand — APPROVED** [review]
Thanks for this, much better than #60042.

Some minor nits which would be nice to clean up, but as this is working as expected, i'm giving +1.

**Comment by carlescufi** [comment]
Let’s address the rest as a follow-up since this is a hotfix

File: cmake/compiler/clang/compiler_flags.cmake (hunk 2 above)
Comment (keith-packard) [reply cmake/compiler/clang/compiler_flags.cmake:27]: I think that's what I did -- the issue here is that clang starts with the gcc settings and modifies them, so I had to reset the property back to the default value.

File: CMakeLists.txt (hunk 3 above)
Comment (keith-packard) [reply CMakeLists.txt:488]: yeah, was pretty unsure about this -- there's a difference between the compiler driver (gcc/clang/cc) and the linker (bfd, ld, lld). This particular flag gets applied to the compiler driver, not directly to the linker?

**Pull request merged**